File: backend/app/schemas/contracts.py
"""Response schemas for contract API routes."""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import (  # noqa: F401  (Field re-export for downstream importers)
    BaseModel,
    ConfigDict,
    Field,
)

from app.schemas.contract_intake import (
    DuplicateContractCandidateResponse,
    ExtractedContractMetadataResponse,
)


class ExtractedFieldResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    field_name: str
    value_json: Any
    span_start: int | None
    span_end: int | None
    span_text: str | None
    confidence: float
    model_name: str
    prompt_version: str
    extracted_at: datetime


class ClauseResponse(BaseModel):
    """Public projection of a persisted Clause.

    Deliberately omits `organization_id`, `embedding`, `created_at`,
    and `updated_at`: the org id is implicit in the contract scope and
    the rest are not surfaced in the v1 UI. Storage / encryption
    fields never appear here because Clause has no such columns.
    """

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    contract_id: uuid.UUID
    ordinal: int
    heading: str | None
    clause_type: str | None
    clause_type_source: str | None
    text: str
    span_start: int
    span_end: int
    confidence: float | None
    segmentation_method: str
    model_name: str | None
    prompt_version: str | None


class ContractListItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    status: str
    mime_type: str
    file_hash_sha256: str
    page_count: int | None
    created_at: datetime
    updated_at: datetime
    # PR #76 — duplicate-merge bookkeeping. ``merged_into_contract_id``
    # is None on canonical records and on records that have never
    # been merged. The default list filters merged rows out at the
    # query layer; these fields exist so a directly-linked merged
    # detail page can still render a safe "merged into …" notice.
    merged_into_contract_id: uuid.UUID | None = None
    merged_at: datetime | None = None
    # PR #101 — when the list endpoint is called with ``?q=…``, each
    # row carries a small hint about *why* the record matched: against
    # its title, its Text preview body, or both. ``None`` when no
    # query is active. The field is deliberately a closed enum of
    # ``"title" | "text_preview" | "title_and_text_preview"`` so the
    # UI never has to interpret raw snapshot content; storage
    # internals, document bytes, private URLs, ``metadata_json``, and
    # any raw ``markdown_text`` are still not returned in the list
    # response.
    search_match_source: str | None = None


class ContractUploadResponse(ContractListItemResponse):
    extracted_fields: list[ExtractedFieldResponse]
    clauses: list[ClauseResponse] = []
    message: str | None = None
    # PR #66 — upload-intake intelligence. Both fields default to "none
    # detected" so older clients that don't render them won't see
    # missing keys; storage internals are excluded by construction at
    # the projection schemas.
    extracted_metadata: ExtractedContractMetadataResponse | None = None
    duplicate_candidates: list[DuplicateContractCandidateResponse] = Field(
        default_factory=list
    )


class ContractDetailResponse(ContractListItemResponse):
    full_text: str | None
    extracted_fields: list[ExtractedFieldResponse]
    clauses: list[ClauseResponse] = []

File: frontend/src/types/contracts.ts
import type {
  DuplicateContractCandidate,
  ExtractedContractMetadata,
} from "./contractIntake";

export type ContractStatus =
  | "uploaded"
  | "extracting"
  | "ready"
  | "failed"
  | "sent_for_signature"
  | "executed";

export interface ContractListItem {
  id: string;
  title: string;
  status: ContractStatus | string;
  mime_type: string;
  file_hash_sha256: string;
  page_count: number | null;
  created_at: string;
  updated_at: string;
  /**
   * PR #76 — when set, this Repository record has been merged into
   * another record (the canonical one). The default list filters
   * these out; detail pages still resolve and render a safe merged
   * notice. Older API responses that don't carry the field treat it
   * as null/absent.
   */
  merged_into_contract_id?: string | null;
  merged_at?: string | null;
}

export type ExtractedFieldValue = unknown;

export interface ExtractedField {
  field_name: string;
  value_json: ExtractedFieldValue;
  span_start: number | null;
  span_end: number | null;
  span_text: string | null;
  confidence: number;
  model_name: string;
  prompt_version: string;
  extracted_at: string;
}

/**
 * One persisted clause from the segmentation pipeline. `text` MUST equal
 * `contract.full_text.slice(span_start, span_end)` exactly. The frontend
 * still re-checks before highlighting, since the backend can lie or rot.
 */
export interface Clause {
  id: string;
  contract_id: string;
  ordinal: number;
  heading: string | null;
  clause_type: string | null;
  clause_type_source: string | null;
  text: string;
  span_start: number;
  span_end: number;
  confidence: number | null;
  segmentation_method: string;
  model_name: string | null;
  prompt_version: string | null;
}

export interface ContractDetail extends ContractListItem {
  full_text: string | null;
  extracted_fields: ExtractedField[];
  clauses: Clause[];
}

export interface UploadContractResponse extends ContractListItem {
  extracted_fields: ExtractedField[];
  clauses: Clause[];
  message?: string | null;
  // PR #66 — upload-intake suggestions + warning-level duplicate
  // candidates. Either may be empty; neither blocks the upload.
  extracted_metadata?: ExtractedContractMetadata | null;
  duplicate_candidates?: DuplicateContractCandidate[];
}

/**
 * A stored file-like artifact tied to a contract. The original upload
 * is recorded as `artifact_type === "original_upload"` with
 * `is_official === true`. Future PRs surface generated DOCX, signed
 * PDFs, redlines, and exhibits as additional artifact rows.
 *
 * The listing endpoint is metadata-only — `storage_key` is intentionally
 * omitted from this type. Use the existing download endpoint to fetch
 * the original artifact's bytes.
 */
export interface ContractArtifact {
  id: string;
  contract_id: string;
  artifact_type: string;
  storage_backend: string;
  filename: string | null;
  mime_type: string | null;
  file_hash_sha256: string | null;
  size_bytes: number | null;
  source: string | null;
  is_official: boolean;
  created_at: string;
  metadata_json: Record<string, unknown> | null;
}

/**
 * A persisted Markdown working snapshot for a contract. The DOCX or
 * PDF remains the original legal artifact; this is a lightweight
 * representation used for fast preview and search.
 */
export interface ContractMarkdownSnapshot {
  id: string;
  contract_id: string;
  markdown_text: string;
  source_kind: string;
  converter_name: string;
  converter_version: string | null;
  conversion_status: string;
  conversion_warnings: unknown[] | null;
  created_at: string;
}
